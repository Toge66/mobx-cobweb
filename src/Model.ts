/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:53. *
 ***************************************************/
import { Model as _Model } from './datx'
import { withStorage } from './mixins/withStorage'
import { withNetActions } from './mixins/withNetActions'

/**** For DTS File (Don`t remove it) ****/
import { INetActionsMixinForModel } from './interfaces/INetActionsMixin'
import { IModelConstructor } from './datx'

export class Model extends withStorage(withNetActions(_Model)) {}
