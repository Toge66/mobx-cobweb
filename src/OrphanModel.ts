/***************************************************
 * Created by nanyuantingfeng on 2020/7/21 11:40. *
 ***************************************************/
import { Model } from './Model'
import { attribute } from './index'
import { ORPHAN_MODEL_ID_KEY, ORPHAN_MODEL_ID_VAL } from './helpers/consts'

export class OrphanModel extends Model {
  static enableAutoId = true

  static preprocess(data: any = {}) {
    data[ORPHAN_MODEL_ID_KEY] = ORPHAN_MODEL_ID_VAL
    return data
  }

  @attribute({ isIdentifier: true })
  public [ORPHAN_MODEL_ID_KEY]: string
}
