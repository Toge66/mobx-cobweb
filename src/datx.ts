/***************************************************
 * Created by nanyuantingfeng on 2020/7/28 09:28. *
 ***************************************************/
export * from '@issues-beta/datx'
