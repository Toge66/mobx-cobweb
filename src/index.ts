/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
export * from 'datx'
export * from './interfaces'
export * from './implements/helpers/model'

export type { ISkeletonCollection } from './interfaces/ISkeletonCollection'
export type { ISkeletonModel } from './interfaces/ISkeletonModel'
export type { ISkeletonView } from './interfaces/ISkeletonView'
export type { IRawResponse } from './interfaces/IRawResponse'
export type { IRequestOptions } from './interfaces/IRequestOptions'
export type { IResponseView } from './interfaces/IResponseView'

export { ResponseView } from './implements/ResponseView'
export { GenericModel } from './implements/GenericModel'
export { GenericView } from './implements/GenericView'
export { setNetworkAdapter } from './implements/NetworkUtils'
export { clearAllCache, clearCacheByType } from './implements/cache'

export { skeleton } from './mixin'
export * from './views'

export * from './implements/adapter/NetworkAdapter'
