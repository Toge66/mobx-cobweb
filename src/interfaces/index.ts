/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 14:12. *
 ***************************************************/
export * from './types'
export * from './IRawResponse'
export * from './IRequestOptions'
export * from './IRawData'
export * from './IResponseHeaders'
export * from './INetworkAdapter'
