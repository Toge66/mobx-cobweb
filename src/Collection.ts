/***************************************************
 * Created by nanyuantingfeng on 2020/6/2 12:55. *
 ***************************************************/
import { Collection as _Collection, IModelConstructor, IType, PureModel, ICollectionConstructor } from 'datx'
import { withNetPatches } from './mixins/withNetPatches'
import { withStorage } from './mixins/withStorage'
import { INetPatchesCollectionMixin } from './interfaces/INetPatchesCollectionMixin'
import { IStorageCollectionMixin } from './interfaces/IStorageCollectionMixin'

const WithNetPatchesCollection: ICollectionConstructor<
  IStorageCollectionMixin<_Collection> & INetPatchesCollectionMixin<_Collection> & _Collection
> = withStorage(withNetPatches(_Collection))

export class Collection extends WithNetPatchesCollection {
  static register<T extends PureModel>(O: IModelConstructor<T>) {
    if (!this.types.find((Q) => Q.type === O.type)) {
      this.types.push(O)
    }
  }

  register<T extends PureModel>(O: IModelConstructor<T>) {
    ;((this.constructor as unknown) as Collection).register(O)
  }

  findOrphan<T extends PureModel>(model?: IType | IModelConstructor<T>): T {
    return this.findOne(model, '__$orphan_id__(0)')
  }
}
