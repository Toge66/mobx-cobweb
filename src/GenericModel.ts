/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import { IModelConstructor, PureModel } from 'datx'
import { withNetActions } from './mixins/withNetActions'
import { INetActionsMixin } from './interfaces/INetActionsMixin'
const WithNetActionModel: IModelConstructor<INetActionsMixin<PureModel> & PureModel> = withNetActions(PureModel)

export class GenericModel extends WithNetActionModel {}
