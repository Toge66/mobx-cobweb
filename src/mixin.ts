/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import {
  ICollectionConstructor,
  IModelConstructor,
  isCollection,
  isModel,
  isView,
  IViewConstructor,
  PureCollection,
  PureModel,
  View
} from 'datx'

import { decorateCollection } from './implements/decorateCollection'
import { decorateModel } from './implements/decorateModel'
import { decorateView } from './implements/decorateView'
import { ISkeletonCollection, ISkeletonModel, ISkeletonModelConstructor } from './interfaces'

export function skeleton<T extends PureModel>(
  BaseClass: IModelConstructor<T>
): ISkeletonModelConstructor<T & ISkeletonModel>

export function skeleton<T extends PureCollection>(
  BaseClass: ICollectionConstructor<T>
): ICollectionConstructor<T & ISkeletonCollection>

export function skeleton<T extends PureModel | PureCollection | View>(
  BaseClass: IModelConstructor<T> | ICollectionConstructor<T> | IViewConstructor<T>
) {
  if (isModel(BaseClass)) {
    return decorateModel(BaseClass as typeof PureModel)
  } else if (isCollection(BaseClass)) {
    return decorateCollection(BaseClass as typeof PureCollection)
  } else if (isView(BaseClass)) {
    return decorateView<T>(BaseClass as typeof View)
  }

  throw new Error('The instance needs to be a model, collection or a view')
}
