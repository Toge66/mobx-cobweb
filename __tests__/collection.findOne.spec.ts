import { collection, useFixturesByGET } from './config'

import { modelToJSON } from '../src'
import Staff from './models/Staff'
import Me from './models/Me'

describe('collection.findOne', () => {
  let scope: any = null

  beforeEach(() => {
    scope = useFixturesByGET()
    collection.removeAll(Staff)
  })

  test('should findOne at collection', async () => {
    const sf = await collection.fetch<Staff>(Staff)

    const staff = collection.findOne<Staff>(Staff, 'cdb28c900c75')
    expect(staff.id).toBe('cdb28c900c75')
    expect(modelToJSON(staff)).toMatchSnapshot()

    const staffx = collection.findOne<Staff>(Staff.type, 'cdb28c900c75')
    expect(staffx.id).toBe('cdb28c900c75')
    expect(staff).toBe(staffx)

    await collection.fetch(Me)
    const me = collection.findAll<Me>(Me)[0]

    expect(modelToJSON(staff)).toMatchSnapshot()
    expect(staff === me.staff).toBeTruthy()

    const staff2 = collection.findOne(Staff, 'cdb28c900c75')
    expect(staff === staff2).toBeTruthy()

    const staff3 = (sf.data as Staff[]).find((s) => s.id === staff.id)

    expect(staff === staff3).toBeTruthy()

    expect(staff.id).toBe('cdb28c900c75')
    expect(modelToJSON(staff)).toMatchSnapshot()
  })
})
