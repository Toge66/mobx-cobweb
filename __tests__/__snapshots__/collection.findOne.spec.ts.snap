// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collection.findOne should findOne at collection 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "6Rk9l1WYNM0400:50128173",
  "departments": Array [
    "6Rk9l1WYNM0400:50128173",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;

exports[`collection.findOne should findOne at collection 2`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "6Rk9l1WYNM0400:50128173",
  "departments": Array [
    "6Rk9l1WYNM0400:50128173",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;

exports[`collection.findOne should findOne at collection 3`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "6Rk9l1WYNM0400:50128173",
  "departments": Array [
    "6Rk9l1WYNM0400:50128173",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;
